File: Yogscart Alpha/Assets/Scripts/Menus/Level_Select.js
#pragma strict

 var currentCup : int;
 var currentTrack : int;
private var gd : CurrentGameData;
private var im : InputManager;
private var mm : MainMenu;

var TypeSelecion : boolean;

var hidden : boolean = true;

var GrandPrixOnly : boolean;

private var Alpha : float;

function Awake(){
	gd = GameObject.Find("GameData").GetComponent(CurrentGameData);
	im = GameObject.Find("GameData").GetComponent(InputManager);

	if(GameObject.Find("Menu Holder").GetComponent(MainMenu) != null)
		mm = GameObject.Find("Menu Holder").GetComponent(MainMenu);
		
}

function OnGUI () {

	var canInput : boolean = true;
	
	if(mm != null && mm.transitioning)
		canInput = false;

if(hidden == false)
Alpha = Mathf.Lerp(Alpha,1,Time.deltaTime*5);
else
Alpha = Mathf.Lerp(Alpha,0,Time.deltaTime*5);

GUI.skin = Resources.Load("GUISkins/Main Menu", GUISkin);
GUI.color = Color(256,256,256,Alpha);

if(GrandPrixOnly)
	TypeSelecion = false;

var submitBool : boolean;
var cancelBool : boolean;
var vert : boolean;
var hori : float;

if(!hidden && canInput){
submitBool = im.c[0].GetMenuInput("Submit") != 0;
cancelBool = im.c[0].GetMenuInput("Cancel") != 0;
vert = im.c[0].GetMenuInput("Vertical") != 0;
hori = im.c[0].GetMenuInput("Horizontal");
}

if(hidden == false){
//Do Input
if(vert && !GrandPrixOnly){
TypeSelecion = !TypeSelecion;
} 

if(cancelBool){
	// Exit Stuff
	if(!Network.isServer && !Network.isClient)
	{
		var mm = GameObject.Find("Menu Holder").GetComponent(MainMenu);
		
		mm.CancelCharacterSelect();
		mm.StartCoroutine("StartCharacterSelect");
		
		hidden = true;
		this.enabled = false;
	}
}

if(TypeSelecion){ //Track Selection

if(hori != 0){
currentTrack += Mathf.Sign(hori);
if(currentTrack < 0)
currentTrack = 3;

if(currentTrack > 3)
currentTrack = 0;
} 

if(submitBool){
Finished();
}

}else{

if(hori){
currentCup += Mathf.Sign(hori);
if(currentCup < 0)
currentCup = gd.Tournaments.Length-1;

if(currentCup >= gd.Tournaments.Length)
currentCup = 0;

}

if(GrandPrixOnly && submitBool){
currentTrack = 0;
Finished();
}

}
}

//Get Textures
var LevelHolder : Texture2D = Resources.Load("UI Textures/Level Selection/LevelHolder",Texture2D);
var Selected : Texture2D = Resources.Load("UI Textures/Level Selection/Selected",Texture2D);
var Tab : Texture2D = Resources.Load("UI Textures/Level Selection/Tab",Texture2D);
var SelectedTab : Texture2D = Resources.Load("UI Textures/Level Selection/SelectedTab",Texture2D);
//Get Width/Ratio/Height of Level Holder
var Width : float = Screen.width-150;
var Ratio : float = Width/LevelHolder.width;
var Height : int = LevelHolder.height * Ratio;

if(!hidden && canInput)
	var click = im.GetClick();

//Render Tracks
for(var j : int = 0; j < gd.Tournaments[currentCup].Tracks.Length; j++){
var OverallRect : Rect = Rect(75 + ((j+1)*14.5f*Ratio) + (j*238f*Ratio),Screen.height/2f + (16f*Ratio),239f*Ratio,210f*Ratio);

GUI.DrawTexture(OverallRect,gd.Tournaments[currentCup].Tracks[j].Logo);

if(TypeSelecion && currentTrack == j)
GUI.DrawTexture(OverallRect,Selected);

if(!hidden && !GrandPrixOnly && im.MouseIntersects(OverallRect) && canInput)
{
currentTrack = j;
TypeSelecion = true;

if(click)
	Finished();

}



}


//Render Tournaments
var TabSize : int = (Width-20-(gd.Tournaments.Length*10))/gd.Tournaments.Length;
TabSize = Mathf.Clamp(TabSize,0,(Width-20-(12*10))/12);

for(var i : int = 0; i < gd.Tournaments.Length; i++){
var TRect : Rect = Rect(85 + (i*10) + (i*TabSize),Screen.height/2f - TabSize,TabSize,TabSize);

if(currentCup == i && !TypeSelecion)
GUI.DrawTexture(TRect,SelectedTab);
else
GUI.DrawTexture(TRect,Tab);

GUI.DrawTexture(TRect,gd.Tournaments[i].Icon);

if(!hidden && im.MouseIntersects(TRect) && canInput)
{
currentCup = i;
TypeSelecion = false;

if(GrandPrixOnly && click)
	Finished();
}

}


//Render level holder
var LHRect : Rect = Rect(75,Screen.height/2,Width,Height);
GUI.DrawTexture(LHRect,LevelHolder);

if(GrandPrixOnly){

var rankText : String = gd.Tournaments[currentCup].LastRank[gd.Difficulty];
var rankRect : Rect = Rect(75,Screen.height/2 + 10 + Height/1.5f,Width,Height);
OutLineLabel(rankRect,rankText,2,Color.black);


}

	if(transform.GetComponent(RaceLeader).type == RaceStyle.TimeTrial)
	{
		var timeRect : Rect = Rect(75,Screen.height/2 + 10 + Height/1.5f,Width,Height);
		var timeString : String = gd.Tournaments[currentCup].Tracks[currentTrack].BestTrackTime.ToString();
		OutLineLabel(timeRect,timeString,2,Color.black);
	}

}


function Finished(){
	if(Network.isServer == true || Network.isClient == true){
		SendRPC();
	}else{
		//Single Player Stuff
		gd.currentCup = currentCup;
		
		if(!GrandPrixOnly)
			gd.currentTrack = currentTrack;
		else
		{
			gd.currentTrack = 0;
		}
		
		this.enabled = false;
		
	}
}

function SendRPC(){
Debug.Log("Send me a RPC");

if(Network.isClient == true)
GetComponent.<NetworkView>().RPC ("LevelChoose",RPCMode.Server,currentCup,currentTrack);
else
transform.GetComponent(RaceLeader).LevelChoose(currentCup,currentTrack);

transform.GetComponent(VotingScreen).hidden = false;
hidden = true;

}


function WithinBounds(Area : Rect){

if(Input.mousePosition.x >= Area.x && Input.mousePosition.x <= Area.x + Area.width 
&&  Screen.height-Input.mousePosition.y >= Area.y &&  Screen.height-Input.mousePosition.y <= Area.y + Area.height)
return true;
else
return false;

}

function OutLineLabel(pos : Rect, text : String,Distance : float,Colour : Color){
Distance = Mathf.Clamp(Distance,1,Mathf.Infinity);

var style = new GUIStyle(GUI.skin.GetStyle("Label"));
style.normal.textColor = Colour;
GUI.Label(Rect(pos.x+Distance,pos.y,pos.width,pos.height),text,style);
GUI.Label(Rect(pos.x,pos.y+Distance,pos.width,pos.height),text,style);
GUI.Label(Rect(pos.x-Distance,pos.y,pos.width,pos.height),text,style);
GUI.Label(Rect(pos.x,pos.y-Distance,pos.width,pos.height),text,style);
var nstyle = new GUIStyle(GUI.skin.GetStyle("Label"));
nstyle.normal.textColor.a = Colour.a;
GUI.Label(pos,text,nstyle);

}

File: Yogscart C#/Assets/Scripts/Menus/LevelSelect.cs
using UnityEngine;
using System.Collections;

public class LevelSelect : MonoBehaviour
{
    CurrentGameData gd;
    public GUISkin skin;
    bool state = false;

    private int currentCup = 0, currentTrack = 0;
    public int trackNum = 4;

    //Transtition
    private float menuAlpha = 1f;
    public bool sliding = false, canClick = false;
    private float[] cupScales, trackScales;

    private SoundManager sm;

    // Use this for initialization
    void Start()
    {
        gd = FindObjectOfType<CurrentGameData>();
        sm = FindObjectOfType<SoundManager>();

        gd.CountGhosts();

        trackScales = new float[] { 1f, 1f, 1f, 1f };
        cupScales = new float[gd.tournaments.Length];

        for (int i = 0; i < cupScales.Length; i++)
            cupScales[i] = 1f;

        if (skin == null)
            skin = Resources.Load<GUISkin>("GUISkins/LevelSelect");

    }

    void OnGUI()
    {
        CurrentGameData.blackOut = false;
        GUI.matrix = GUIHelper.GetMatrix();
        GUI.skin = skin;
        GUI.depth = -5;

        canClick = false;

        GUIHelper.SetGUIAlpha(menuAlpha);
        Vector2 mousePos = GUIHelper.GetMousePosition(); 

        float cupNameHeight = GUIHelper.height / 2f;
        float individualHeight = cupNameHeight / Mathf.Clamp(gd.tournaments.Length, 4, 8);
        Texture2D rectangle = Resources.Load<Texture2D>("UI/Level Selection/Rectangle");
        float startY = 200;

        //Draw Tournaments
        for (int i = 0; i < gd.tournaments.Length; i++)
        {
            Rect rectRect = GUIHelper.CentreRect(new Rect(300, startY + (i * individualHeight) + 10, 400, individualHeight - 20), cupScales[i]);
            GUI.DrawTexture(rectRect, rectangle);

            //Scale the Cup Names
            bool contains = Cursor.visible && rectRect.Contains(mousePos);
            if (!state && (contains || currentCup == i))
            {
                if (cupScales[i] < 1.1f)
                    cupScales[i] += Time.unscaledDeltaTime;
                else
                    cupScales[i] = 1.1f;

                if (contains)
                {
                    currentCup = i;
                    canClick = true;
                }
            }
            else
            {
                if (cupScales[i] > 1f)
                    cupScales[i] -= Time.unscaledDeltaTime;
                else
                    cupScales[i] = 1f;
            }

            GUIHelper.CentreRectLabel(new Rect(300, startY + (i * individualHeight) + 15, 400, individualHeight - 30), cupScales[i], gd.tournaments[i].name, (currentCup == i)?Color.yellow:Color.white);
        }

        Race gamemode = (Race)CurrentGameData.currentGamemode;

        int tempCurrentCup = Mathf.Clamp(currentCup, 0, 4);

        //Draw Tracks
        for (int i = 0; i < gd.tournaments[tempCurrentCup].tracks.Length; i++)
        {
            Texture2D trackPreview;

            if ((gamemode is TimeTrial || gamemode is VSRace) && state && i != currentTrack)
            {
                trackPreview = gd.tournaments[tempCurrentCup].tracks[i].logo_GreyOut;
            }
            else
            {
                trackPreview = gd.tournaments[tempCurrentCup].tracks[i].logo;
            }

            Rect previewRect = GUIHelper.CentreRect(new Rect(800 + (500 * (i % 2)), startY - 120 + (390 * (i / 2)), 400, 365), trackScales[i]);

            //Scale the Track Names
            bool contains = Cursor.visible && previewRect.Contains(mousePos);
            if (state && (contains || currentTrack == i))
            {
                if (trackScales[i] < 1.1f)
                    trackScales[i] += Time.unscaledDeltaTime;
                else
                    trackScales[i] = 1.1f;

                if (contains)
                {
                    currentTrack = i;
                    canClick = true;
                }
            }
            else
            {
                if (trackScales[i] > 1f)
                    trackScales[i] -= Time.unscaledDeltaTime;
                else
                    trackScales[i] = 1f;
            }

            GUI.DrawTexture(previewRect, trackPreview, ScaleMode.ScaleToFit);
        }

        Rect timeRect = new Rect(810, startY + 640, 1000, 100);
        Rect ghostRect = new Rect(810, startY + 690, 1000, 100);

        if (state && gamemode is TimeTrial && gd.tournaments[tempCurrentCup].tracks.Length > currentTrack && currentTrack != -1)
        {
            string timeString = "";
            if (gd.tournaments[tempCurrentCup].tracks[currentTrack].bestTime != 0f)
                timeString = TimeManager.ToString(gd.tournaments[tempCurrentCup].tracks[currentTrack].bestTime);
            else
                timeString = "N/A";

            GUIHelper.OutLineLabel(timeRect, "Best Time:  " + timeString, 2, Color.black);

            string ghostString = gd.tournaments[tempCurrentCup].tracks[currentTrack].ghosts.ToString();
            GUIHelper.OutLineLabel(ghostRect, "Local Ghosts:  " + ghostString, 2, Color.black);
        }
        else if(gamemode is TournamentRace && !(gamemode is VSRace))
        {
            string rank = gd.tournaments[tempCurrentCup].lastRank[CurrentGameData.difficulty].ToString();
            if (rank == "NoRank")
                rank = "No Rank";

            GUIHelper.OutLineLabel(timeRect, rank, 2, Color.black);
        }
        else if(gamemode is VSRace)
        {
            if (GetComponent<MainMenu>() != null)
            {
                //Draw Inputs
                Rect leftRect = new Rect(1000, startY + 650, 160, 80);
                Rect rightRect = new Rect(1450, startY + 650, 160, 80);

                GUI.DrawTexture(leftRect, InputManager.controllers.Count > 0 && InputManager.controllers[0].inputType == InputType.Xbox360 ? Resources.Load<Texture2D>("UI/Options/LB") : Resources.Load<Texture2D>("UI/Options/Q"));
                GUI.DrawTexture(rightRect, InputManager.controllers.Count > 0 && InputManager.controllers[0].inputType == InputType.Xbox360 ? Resources.Load<Texture2D>("UI/Options/RB") : Resources.Load<Texture2D>("UI/Options/E"));

                GUIStyle customButton = new GUIStyle();

                if(GUI.Button(leftRect, "", customButton))
                {
                    trackNum = MathHelper.NumClamp(trackNum - 1, 1, 64);
                }

                if (GUI.Button(rightRect, "", customButton))
                {
                    trackNum = MathHelper.NumClamp(trackNum + 1, 1, 64);
                }

                GUIHelper.OutLineLabel(timeRect, "Races: " + trackNum, 2, Color.black);
            }
            else
            {
                VSRace vs = gamemode as VSRace;
                GUIHelper.OutLineLabel(timeRect, "Race " + (gamemode.currentRace + 1).ToString() + " / " + vs.raceCount, 2, Color.black);
            }
        }

        //Inputs
        int vert = 0, hori = 0, tabChange = 0;
        bool submit = false, cancel = false;

        if (!sliding && (FindObjectOfType<MainMenu>() == null || !FindObjectOfType<MainMenu>().sliding))
        {
            vert = InputManager.controllers[0].GetIntInputWithLock("MenuVertical");
            hori = InputManager.controllers[0].GetIntInputWithLock("MenuHorizontal");
            submit = InputManager.controllers[0].GetButtonWithLock("Submit");
            cancel = InputManager.controllers[0].GetButtonWithLock("Cancel");
            tabChange = InputManager.controllers[0].GetIntInputWithLock("TabChange");

            if (canClick && Input.GetMouseButton(0))
                submit = true;

            if (hori != 0)
            {
                if (state)
                {
                    if (currentTrack <= 1)
                        currentTrack = MathHelper.NumClamp(currentTrack + hori, 0, 2);
                    else
                        currentTrack = MathHelper.NumClamp(currentTrack + hori, 2, 4);

                    hori = 0;
                }
            }

            if (vert != 0)
            {
                if (state)
                    currentTrack += (2 * vert);
                else
                {
                    currentCup += vert;
                    currentTrack = 0;
                }

                vert = 0;
            }

            currentTrack = MathHelper.NumClamp(currentTrack, 0, 4);
            currentCup = MathHelper.NumClamp(currentCup, 0, gd.tournaments.Length);

            if(tabChange != 0 && gamemode is VSRace && GetComponent<MainMenu>() != null)
            {
                trackNum = MathHelper.NumClamp(trackNum + tabChange, 1, 64);
                tabChange = 0;
            }
        }

        //if (gamemode is OnlineRace)
       // {
       //     Race.currentTrack = currentTrack;
       //     Race.currentCup = currentCup;
       // }

        if (submit)
        {
            if (gamemode is TournamentRace && !(gamemode is VSRace))
            {
                FinishLevelSelect();
            }
            else
            {
                if (!state)
                    state = true;
                else
                    FinishLevelSelect();
            }

            sm.PlaySFX(Resources.Load<AudioClip>("Music & Sounds/SFX/confirm"));
            submit = false;
        }

        if (cancel)
        {
            DoCancel();
            cancel = false;
        }

        GUI.color = Color.white;

    }

    public void DoCancel()
    {
        if (FindObjectOfType<MainMenu>() == null || state)
            sm.PlaySFX(Resources.Load<AudioClip>("Music & Sounds/SFX/back"));

        if (!state)
        {
            CancelLevelSelect();
        }
        else
        {
            state = false;
        }
    }

    public void CancelLevelSelect()
    {
        //Cancel out of Level Select
        StartCoroutine(HideLevelSelect());
        currentCup = -1;
        currentTrack = -1;

        GameMode gameMode = CurrentGameData.currentGamemode;

        //If backing out of Level Select make sure Voting Screen is showed
        /*if (gamemode.raceType == RaceType.Online)
        {
            if (FindObjectOfType<VotingScreen>() == null)
                gd.gameObject.AddComponent<VotingScreen>();

            FindObjectOfType<VotingScreen>().ShowScreen();
        }*/

        if(FindObjectOfType<Replay>() != null)
        {
            FindObjectOfType<Replay>().GoBack();
        }
        else if (gameMode is VSRace)
        {
            VSRace race = gameMode as VSRace;
            race.CancelLevelSelect();
        }
    }

    public void ShowLevelSelect()
    {
        currentCup = 0;
        currentTrack = 0;
        trackNum = 4;

        if (FindObjectOfType<MainMenu>() != null && FindObjectOfType<MainMenu>().state != MainMenu.MenuState.LevelSelect)
            FindObjectOfType<MainMenu>().ChangeMenu(MainMenu.MenuState.LevelSelect);

        StartCoroutine(ActualShowLevelSelect());
    }
    private IEnumerator ActualShowLevelSelect()
    {
        sliding = true;

        float startTime = Time.realtimeSinceStartup;
        float travelTime = 0.25f;
        //Slide Off //////////////////////////////
        menuAlpha = 0f;

        while (Time.realtimeSinceStartup - startTime < travelTime)
        {
            menuAlpha = Mathf.Lerp(0f, 1f, (Time.realtimeSinceStartup - startTime) / travelTime);
            yield return null;
        }

        menuAlpha = 1f;

        sliding = false;
    }

    private void FinishLevelSelect()
    {
        if (Time.timeScale != 1f)
            Time.timeScale = 1f;

        PauseMenu pm = FindObjectOfType<PauseMenu>();
        if (pm != null)
        {
            pm.pauseHold = -1;
            pm.paused = -1;
            PauseMenu.canPause = false;
        }

        Race gameMode = CurrentGameData.currentGamemode as Race;
        /*if (gamemode.raceType == RaceType.Online)
        {
            //Send data to server
            FindObjectOfType<NetworkRaceClient>().SendVote(currentCup, currentTrack);

            if(FindObjectOfType<VotingScreen>() == null)
                gd.gameObject.AddComponent<VotingScreen>();

            FindObjectOfType<VotingScreen>().ShowScreen();
        }*/

        gameMode.FinishLevelSelect(currentCup, currentTrack);
        StartCoroutine(HideLevelSelect());
    }

    public void ForceFinishLevelSelect()
    {
        StartCoroutine(HideLevelSelect());
    }

    private IEnumerator HideLevelSelect()
    {
        sliding = true;

        float startTime = Time.realtimeSinceStartup;
        float travelTime = 0.25f;
        //Slide Off //////////////////////////////
        menuAlpha = 1f;

        while (Time.realtimeSinceStartup - startTime < travelTime)
        {
            menuAlpha = Mathf.Lerp(1f, 0f, (Time.realtimeSinceStartup - startTime) / travelTime);
            yield return null;
        }

        menuAlpha = 0f;

        sliding = false;
        enabled = false;
    }

    private IEnumerator KillLevelSelect()
    {
        yield return HideLevelSelect();
        Destroy(this);
    }
}